File: backend/app/services/annotation_service.py
from sqlalchemy.orm import Session
from fastapi import HTTPException

from app.repositories.annotation_repository import AnnotationRepository
from app.models.user import User
import json


class AnnotationService:
    def __init__(self, db: Session):
        self.repo = AnnotationRepository(db)

    def list_by_document(self, document_id: int, user: User) -> list[dict]:
        annotations = self.repo.list_by_document(document_id)
        # Filter by user ownership
        result = []
        for ann in annotations:
            if ann.user_id != user.id:
                continue
            result.append(self._to_dict(ann))
        return result

    def create(self, data: dict, user: User) -> dict:
        data["document_id"] = data.get("document_id")
        ann = self.repo.create(user.id, data)
        return self._to_dict(ann)

    def delete(self, annotation_id: int, user: User):
        ann = self.repo.get_by_id(annotation_id)
        if not ann or ann.user_id != user.id:
            raise HTTPException(status_code=404, detail="Annotation not found")
        self.repo.soft_delete(annotation_id)

    def _to_dict(self, ann) -> dict:
        pos_data = {}
        try:
            pos_data = json.loads(ann.position_data) if ann.position_data else {}
        except (json.JSONDecodeError, TypeError):
            pos_data = {}
        return {
            "id": ann.id,
            "document_id": ann.document_id,
            "page": ann.page_number or 1,
            "selected_text": ann.selected_text or "",
            "color": ann.color or "#FFD700",
            "type": ann.annotation_type or "highlight",
            "note": ann.note or "",
            "position_data": pos_data,
            "created_at": ann.created_at.isoformat() if ann.created_at else "",
        }

File: backend-go/internal/service/annotation_service.go
package service

import (
	"encoding/json"
	"errors"
	"time"

	"github.com/chenxxianyi/NoteWeb/backend-go/internal/models"
)

type AnnotationService struct {
	repo AnnotationRepository
}

type AnnotationRepository interface {
	GetByID(id uint) (*models.Annotation, error)
	ListByDocument(docID uint) ([]models.Annotation, error)
	Create(ann *models.Annotation) error
	SoftDelete(id uint) error
	Replace(
		userID uint,
		documentID uint,
		deleteIDs []uint,
		creates []models.Annotation,
	) ([]models.Annotation, error)
}

func NewAnnotationService(repo AnnotationRepository) *AnnotationService {
	return &AnnotationService{repo: repo}
}

type AnnotationResponse struct {
	ID           uint                   `json:"id"`
	DocumentID   uint                   `json:"document_id"`
	Page         int                    `json:"page"`
	SelectedText string                 `json:"selected_text"`
	Color        string                 `json:"color"`
	Type         string                 `json:"type"`
	Note         string                 `json:"note"`
	PositionData map[string]interface{} `json:"position_data"`
	CreatedAt    string                 `json:"created_at"`
}

func toAnnResponse(a *models.Annotation) AnnotationResponse {
	posData := make(map[string]interface{})
	json.Unmarshal([]byte(a.PositionData), &posData)
	return AnnotationResponse{
		ID: a.ID, DocumentID: a.DocumentID, Page: a.PageNumber,
		SelectedText: a.SelectedText, Color: a.Color, Type: a.AnnotationType,
		Note: a.Note, PositionData: posData,
		CreatedAt: a.CreatedAt.Format(time.RFC3339),
	}
}

func (s *AnnotationService) ListByDocument(docID, userID uint) ([]AnnotationResponse, error) {
	anns, err := s.repo.ListByDocument(docID)
	if err != nil {
		return nil, err
	}
	result := make([]AnnotationResponse, 0)
	for _, a := range anns {
		if a.UserID == userID {
			result = append(result, toAnnResponse(&a))
		}
	}
	return result, nil
}

func (s *AnnotationService) Create(userID uint, req AnnotationCreateRequest) (*AnnotationResponse, error) {
	posBytes, _ := json.Marshal(req.PositionData)
	ann := &models.Annotation{
		UserID: userID, DocumentID: req.DocumentID, PageNumber: req.Page,
		SelectedText: req.SelectedText, Color: req.Color,
		AnnotationType: req.AnnType, Note: req.Note,
		PositionData: string(posBytes),
	}
	if err := s.repo.Create(ann); err != nil {
		return nil, err
	}
	resp := toAnnResponse(ann)
	return &resp, nil
}

func (s *AnnotationService) Delete(annID, userID uint) error {
	ann, err := s.repo.GetByID(annID)
	if err != nil || ann.UserID != userID {
		return errors.New("批注不存在")
	}
	return s.repo.SoftDelete(annID)
}

func (s *AnnotationService) Replace(
	userID uint,
	req AnnotationReplaceRequest,
) ([]AnnotationResponse, error) {
	creates := make([]models.Annotation, 0, len(req.Creates))
	for _, create := range req.Creates {
		positionData, err := json.Marshal(create.PositionData)
		if err != nil {
			return nil, err
		}
		creates = append(creates, models.Annotation{
			UserID:         userID,
			DocumentID:     req.DocumentID,
			PageNumber:     create.Page,
			SelectedText:   create.SelectedText,
			Color:          create.Color,
			AnnotationType: create.AnnType,
			Note:           create.Note,
			PositionData:   string(positionData),
		})
	}

	created, err := s.repo.Replace(userID, req.DocumentID, req.DeleteIDs, creates)
	if err != nil {
		return nil, err
	}

	result := make([]AnnotationResponse, 0, len(created))
	for index := range created {
		result = append(result, toAnnResponse(&created[index]))
	}
	return result, nil
}

type AnnotationCreateRequest struct {
	DocumentID   uint                   `json:"document_id"`
	Page         int                    `json:"page"`
	SelectedText string                 `json:"selected_text"`
	Color        string                 `json:"color"`
	AnnType      string                 `json:"type"`
	Note         string                 `json:"note"`
	PositionData map[string]interface{} `json:"position_data"`
}

type AnnotationReplacementCreateRequest struct {
	Page         int                    `json:"page"`
	SelectedText string                 `json:"selected_text"`
	Color        string                 `json:"color"`
	AnnType      string                 `json:"type"`
	Note         string                 `json:"note"`
	PositionData map[string]interface{} `json:"position_data"`
}

type AnnotationReplaceRequest struct {
	DocumentID uint                                 `json:"document_id"`
	DeleteIDs  []uint                               `json:"delete_ids"`
	Creates    []AnnotationReplacementCreateRequest `json:"creates"`
}
